### ISSUE TITLE
Blackmagic Outputs

### ISSUE THREAD

**Issue by webnerdnick** [vassbo/freeshow#124]
You probably already know this, but having the ability use Blackmagic products that can do SDI out would give you a huge bump in market share.  Have you yet investigated what it would take to use an output like a Blackmagic decklink?

**Comment by vassbo** [comment]
It would definitely be a handy feature. But maybe a bit time consuming to add, and I don't have any device to test it with. But I'll see what I can figure out. 😊

**Comment by webnerdnick** [comment]
I saw this project on github: https://github.com/Streampunk/macadam which uses node.js and the Blackmagic API and has a handful of other repositories that it's used in.  I'm not proficient in Electron, or I'd give it a shot myself.  I have some spare decklink PCI cards which if you're interested, I may be able to send you one to play with if you think this is something you'd be interested digging into.  I realize though I'm sure you have a ton of other priorities for this project.

**Comment by vassbo** [comment]
Thank you! Yes, I have some other priorities first. But I'll let you know if I need it, and have the time to do it. The package looks promising.

**Comment by kol0** [comment]
Big +1 to add true SDI output.

**Comment by marcorychlik** [comment]
Big +1 to add true SDI output. 

I was looking for years to a alternative to PP7... i found hope. I pray for you guys for shure. Thanks for our effort. When you @vassbo (and others) ever will come to Switzerland, let me know. I created a git profile to write this.

**Comment by vassbo** [comment]
Thank you. @marcorychlik 😊

**Comment by marcorychlik** [comment]
We thank you!!! looking forward


Marco Rychlik
ICF Zürich - Bereichsleiter Video

Zürichstrasse 131, CH-8600 Dübendorf | ***@***.***
Tel: +41 43 366 76 76 | Direkt: +41 43 366 76 24
Handy: +41 79 656 82 35 | Fax: +41 43 366 76 77

www.icf.ch | «ICF - Kirche neu erleben»




Kristoffer Vassbø ***@***.***> schrieb am So. 7. Apr. 2024 um
22:07:

**Comment by mellbergsimon** [comment]
+1 for me aswell. Would be the incentive for me to switch. Key + Fill is essential in my application.

**Comment by Qwiko** [comment]
+1

**Comment by djrenault** [comment]
Created an account just to say this feature is holding me back from being able to switch as well. Now that ProPresenter have announced a subscription-only model, hopefully this brings in donors who are switching away from Pro7 and fund the features we need to be able to switch and donate too.

**Comment by webnerdnick** [comment]
I was the one that originally requested this feature. I've found the best way of getting around this is to use a regular graphics card and then HDMI to SDI converters. It's basically what a deck link is already doing (minus the input abilities). It's actually easier to set up and troubleshoot going this route. What I do wish was possible would be to lock the mouse or other content from being accidentally moved to one of these screens since windows now has direct access.

**Comment by mellbergsimon** [comment]
> I was the one that originally requested this feature. I've found the best way of getting around this is to use a regular graphics card and then HDMI to SDI converters. It's basically what a deck link is already doing (minus the input abilities). It's actually easier to set up and troubleshoot going this route. What I do wish was possible would be to lock the mouse or other content from being accidentally moved to one of these screens since windows now has direct access. 

I think the main thing decklinks would give is key+fill. Atleast I and I think alot more are using ATEMs or similar to put text on top of other content. That would have to be done with a greenscreen. But that feels like a step in the wrong direction for me. But for simpler playouts where Freeshow would display on the entire screen your suggestion is sufficient.

**Comment by vassbo** [comment]
If you have a Blackmagic device (e.g. DeckLink), and some coding knowledge, please help with testing! :)

I have added basic support for Blackmagic outputs/inputs on this GitHub branch (with the macadam package): https://github.com/ChurchApps/FreeShow/tree/blackmagic

All the basic functionality should be there, like sending out video data (in Outputs settings) & receiving video streams (in "Media">"Screens">"Blackmagic"). But I do not own any Blackmagic device so I can't test it, I have no idea if it works at all!

**Comment by brettgoodrich** [comment]
+1 from me. This is basically The Thing I would need, and that's it.
OH WOW A BETA YESTERDAY
Will try it

**Comment by brettgoodrich** [comment]
@vassbo Do you have a PO Box one of us could send one to? They're $145 for all you would need to test it
https://doddcamera.com/blackmagic-design-ultrastudio-mini-recorder-3g-sdi-hdmi-to-thunderbolt

**Comment by vassbo** [comment]
@brettgoodrich Thanks, I have a friend currently testing it, do you also get a green output screen? If I can't figure out that issue I might have to buy a DeckLink, but that could come in handy elsewhere. Do you think I just need a DeckLink, or would I need some other hardware as well?

**Comment by webnerdnick** [comment]
(I think FreeShow needs a Discord for the purpose of community of engagement, troubleshooting, and chat threads like this, but anyways...) A mini monitor or mini recorder would get you in the ballpark for testing, but for those above who were talking about key+fill, you'd need to get into the PCI DeckLink product line, starting with the DeckLink Duo 2 or DeckLink Quad or something like that (those are older versions), but will certainly allow you to test everything, especially as more features are requested if this SDI out is successful. I'm happy to chip in some money toward this if you need to purchase some additional hardware.

**Comment by Velikyi** [comment]
> I tested [this build](https://github.com/ChurchApps/FreeShow/issues/124#issuecomment-2268629203) with my 8K Pro pipeline. In our setup, we use ProPresenter on three outputs which all go to our ATEM. We normally run these at 1080p59.94 with no transparency (key/fill). Our 4th port is used as 2160p59.94 input from the ATEM's program feed, and that goes into OBS for our livestreaming.
> 
> In this test I just slotted FreeShow into one of the ProPresenter output feeds. I had no visible input at all, until changing the Pixel Format *after* enabling the BMD output. Once changed, both 8-bit YUV and 10-bit YUV seem to 'work', in the sense that they output picture. *What* they output is obviously not what's intended, but it's also *different* depending on the setting.
> 
> Also, the resolution setting doesn't seem to do anything. When changing resolution it does seem to 'refresh' the image (cycling black then back to video), but regardless of whatever I have it set to, it only outputs at 1080i50 according to the BMD Desktop Video Setup software.
> 
> I've attached four images - the outputs of 8-bit and 10-bit YUV, as captured in OBS, when OBS and the ATEM are set to both 4K and 1080p. Hopefully this mostly 'raw' pixel dump might help you troubleshoot what's going on.
> 
> [8bit_1080p](https://github.com/user-attachments/assets/900f62fe-d6d2-4d87-a427-954cf248a505)
> [10bit_1080p](https://github.com/user-attachments/assets/43d1d590-9bb6-48ed-9373-24db56b0bfe5)
> [8bit_4K](https://github.com/user-attachments/assets/04743bf3-7713-42b6-b06c-f6b15f5f129d)
> [10bit_4K](https://github.com/user-attachments/assets/c6703a36-3050-4124-914a-f07a910b9fcd)
> 
> For what it's worth, you do seem to be accurately obtaining the supported video modes for the specific Decklink card, as I see every resolution mode from 525i59.94 NTSC to 4K DCI 60p.

This is exactly what I got too, though with 1080p50 and a mixture of one Quad card and one 8K card

**Comment by vassbo** [comment]
Thank you @Yttrium-tYcLief I'll see if I'm able to do something about this, and post a new test build.

**Comment by vassbo** [comment]
@Yttrium-tYcLief Will let you know, had to prioritize some optimizations and bugfixes first.

**Comment by LPChurchNeeds** [comment]
Any progress on using Black Magic to run Free Show?  Our church is really in a pickle trying to figure this out.  Thanks

**Comment by vassbo** [comment]
@Yttrium-tYcLief I might have figured something out!
Changed the YUV subsampling format from 444 to 420.

Here is a new test build for anyone interested.
https://filetransfer.io/data-package/eqhdTAIU

**Comment by discodancepant** [comment]
I tested the latest (FreeShow-1.2.9-beta.1-x64) with a Decklink Quad 2:
Decklink config:
Output 1: 1080p60 with alpha
Output 2: 1080p60 no alpha

ProPresenter works fine. I included the Decklink Status Report for the two outputs. ProPresenter always uses YUV 4:2:2 8 bit for non-alpha output and BGRA 8 bit for outputs with an alpha.
![Blackmagic Design System Report - DeckLink Quad Output 1](https://github.com/user-attachments/assets/9ed9f0c2-f0eb-4126-9b08-56397424c74c)
![Blackmagic Design System Report - DeckLink Quad Output 2](https://github.com/user-attachments/assets/3c262a00-e6fc-4dd5-b7ea-a3ca4c85ef56)

The Freeshow beta seems to temporarily connect to the decklink once, but if you close and open it, it doesn't detect the Decklink at all. System restart does the same thing, it doesn't detect the Decklink. To re-detect the Decklink I had to remove Freeshow, remove the Freeshow MyDocuments folder, and also remove the folder from the hidden AppData folder. 

From the dropdown, Freeshow in 1080p60 only allows YUV 8 bit and 10 bit, no BGRA.
I configured Freeshow to use 720p60 with BGRA with alpha output, but same green output as always.

I believe the connection issue to the Decklink card needs to be resolved, as it seems like Freeshow momentarily connects, then quickly loses the connection. This might also just be my experience with the Quad 2. I have a Duo 2 as well, but I lacked the time to test it.

**Comment by chalamministries** [comment]
I have a mac and I have a decklink quad and Im proficient in programming. I've dabbled a little with electron. I'll try to give it a shot.

**Comment by vassbo** [comment]
@chalamministries Let me know if you make any progress on this. :)

**Comment by chalamministries** [comment]
I couldn't get it to compile on my main Mac. I got to dig and find out what's going on there. But yeah I'm working on it.

**Comment by LPChurchNeeds** [comment]
Awesome. Thank you!
    On Tuesday, November 12, 2024 at 02:17:04 AM CST, chalamministries ***@***.***> wrote:  
 
 


I couldn't get it to compile on my main Mac. I got to dig and find out what's going on there. But yeah I'm working on it.

**Comment by chalamministries** [comment]
I'm running into a lot of depricated dependencies. Im afraid to upgrade all of this as it would affect the main version. I can't get it to build properly. Am I missing something that you are doing to run it?

**Comment by vassbo** [comment]
All right, do you think this can fix the deprecated dependencies?: https://github.com/ChurchApps/FreeShow/pull/1016

**Comment by vassbo** [comment]
@LPChurchNeeds If you need Blackmagic integration at the moment your best bet is ProPresenter yes if you have the money, but it's expensive. I'm sure we will support devices like DeckLink very soon!

**Comment by LPChurchNeeds** [comment]
Thanks. Yes. We got stuck, couldn't figure it out, seemed like Freestone would work, but it did not. 
    On Thursday, November 14, 2024 at 08:34:37 AM CST, Kristoffer Vassbø ***@***.***> wrote:  
 
 


@LPChurchNeeds If you need Blackmagic integration at the moment your best bet is ProPresenter yes if you have the money, but it's expensive. I'm sure we will support devices like DeckLink very soon!

**Comment by amudhanr** [comment]
@vassbo Thank you for working on this. Will FreeShow eventually support fill/key output for BMD UltraStudio HD Mini or UltraStudio Mini 4K via USB-C output? We are considering buying this hardware for overlays using ATEM SDI Extreme and FreeShow, but the fill/key output is essential for this to work as we stream straight out of the ATEM and not use OBS for overlays. I know we can use overlays using NDI source and OBS, but we are trying to not to use OBS for our streams.

**Comment by vassbo** [comment]
@amudhanr We should have support for BMD key/fill once we get this actually working, not sure when exactly.

**Comment by chalamministries** [comment]
Just letting everyone know, I'm still working on this enhancement. Feel free to build the blackmagic repo and try the outputs. I still have to work on the inputs.

**Comment by chalamministries** [comment]
Hey there. I am working on that branch to optimize the memory consumption. What mode do you want/need to run it in?

[Steven Gauerke](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=2x8x25)	[2x8x25]Web Developer

On December 29, 2024 at 16:57 GMT, Yttrium tYcLief via Info ***@***.***> wrote:

I've built the blackmagic branch as of [ff124cb](https://github.com/ChurchApps/FreeShow/commit/ff124cb2b962fabe43fd7e781ad0c8d7de0e5e1a). The first time I launched it, I had working output on my 8k Pro when set to 1080p59.94 and 8-bit YUV. However, performance was abysmal; while the Freeshow app itself felt fine, and the preview showed a buttery smooth output, the actual render output seemed more like 5fps, if that. There were also a lot of errors in console. Unfortunately, I can't recreate that situation, because while typing up a response here my computer actually bluescreened (IRQL error, suggesting driver issue; this happens with Decklink cards from time to time when run in janky setups). After rebooting, I can't get output anymore, and started getting these nasty error popups:

[image.png (view on web)](https://github.com/user-attachments/assets/98abe977-ad0a-423b-8781-41e184912c0e)

Worth noting that only 8bit YUV seems to work correctly, all of the other modes send a very distorted image. My ATEM does display it in the multiview, so the signal is at least recognizable enough to display, but the image looks very wrong, in different ways in each mode.

**Comment by chalamministries** [comment]
Yup. It's worse when you have moving images. What is your output display? In my tests, my actual output through the blackmagic device was like 3000 something and I was running 1920x1080. So even if you have it set to a smaller resolution, it's still having to scale for that device to display properly.

[Steven Gauerke](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=2x8yfb)	[2x8yfb]Web Developer

On December 29, 2024 at 17:28 GMT, Yttrium tYcLief via Info ***@***.***> wrote:

8-bit YUV at 1080p59.94 is absolutely fine, assuming it performs properly. Was just trying to help by mentioning it for debug reasons.

Memory is certainly an issue (currently consuming 7.5GB and climbing while doing absolutely nothing) but CPU usage is a much bigger deal. This is in debug mode (so I get the electron frame UI) but 30% CPU usage on a Ryzen 5600X, climbing to 50% when displaying a still image, is absolutely unusable.

As I've been typing this response, the memory leak has grown worse... up to 15.2GB now, before Freeshow fully crashed with the following:
[image.png (view on web)](https://github.com/user-attachments/assets/6b51433d-8649-4d7a-b2ac-0711be44462f)

**Comment by taenologue** [comment]
Could it be due to a version issue with the Blackmagic API included in the Macadam library?

**Comment by chalamministries** [comment]
> Could it be due to a version issue with the Blackmagic API included in the Macadam library?

Yes thats what I've been working on. Im either going to have to modify the macadam library or we try and go with an old version of Video Desktop API. 

I only get 1 day a week to work on this so bare with me but it IS coming along. I had successful video output and scaling but then when you played video....that caused issues without having to eat up a ton of memory.

**Comment by josuabarron** [comment]
That’s so awesome, can we push the inputs to the main release or do you want to have it all finished before doing that?

**Comment by DavidRaj23** [comment]
Hi, Can i use FreeShow output with Blackmagic UltraStudio Monitor 3G now with this latest build or still this in dev progress ?

**Comment by vassbo** [comment]
@DavidRaj23 Not directly from the program yet.

**Comment by avanere** [comment]
Hi, While the development is still in progress, have anyone tried using https://github.com/norihiro/obs-output-filter ? Freeshow -> display -> OBS (obs-decklink-output-filter) -> Decklink SDI. 
I have not tested it as I don't have any Decklink but if somebody can test and confirm that this works, then this can be used as a temporarily solution. :)

**Comment by soundguybob** [comment]
Any chance we can also get support for AJA cards? I much prefer their hardware over BMD, as I’ve had too many failures with various BMD products over the years.

**Comment by chalamministries** [comment]
The struggle is that we need some of that hardware to be able to test with. I have a blackmagic decklink card but no enclosure. Nor do I have AJA devices. If I Can get my hands on one, I'll see what I can do.

**Comment by discodancepant** [comment]
> The struggle is that we need some of that hardware to be able to test with. I have a blackmagic decklink card but no enclosure. Nor do I have AJA devices. If I Can get my hands on one, I'll see what I can do.

I have two decklinks, A Blackmagic Decklink Duo 2 and a Decklink Quad 2. I can't spare them, but I'm willing to test builds as needed, but I need someone to let me know there are changes to Decklink output capability. I'm a software engineer myself, so I can easily follow a list of steps if you'd like me to test each (relevant) build, if there any debugging logs I can post with results.

**Comment by avanere** [comment]
We were able to test OBS Decklink Output filter  (https://github.com/cg2121/obs-decklink-output-filter) today with the AV team at our church. We were able to show three scenes at the same time going to Decklink SDI output 1,3 and 4. So, it's possible for those who are currently using Blackmagic Decklink to use it with Freeshow. It would be better to wait for SDI output feature in Freshow but if needed, you can use OBS for now. :)

**Comment by bfranske** [comment]
I have a system with a decklink in it that I could make available for testing long-term via VPN if that would be helpful. I use the system for about 4 weeks a year (late april-mid-may) but the rest of the year it is available.

**Comment by chalamministries** [comment]
I'd be interested. That's what I was hoping for. Are you on the slack channel? Steven Gauerke is my screenname.

**Comment by vassbo** [comment]
Bought a Macbook with Thunderbolt and a Blackmagic UltraStudio Monitor to debug the color issues, I think it should be working now: https://drive.google.com/file/d/1m6fkVkCxgIsgpmguMO6tHuIb7WnYLM1Z/view

**Comment by Yttrium-tYcLief** [comment]
> Bought a Macbook with Thunderbolt and a Blackmagic UltraStudio Monitor to debug the color issues, I think it should be working now: https://drive.google.com/file/d/1m6fkVkCxgIsgpmguMO6tHuIb7WnYLM1Z/view

I've been getting quite mad with our ProPresenter 7.13 instance crashing so frequently lately, so I decided to take another look and see what progress had been made here. I just tried out this build and it seems to perform great and work pretty well in the default 720p60 in 8-bit YUV mode. However, I can't get any other resolutions to work on my Decklink 8K Pro on Win11. ~~The NTSC/PAL (lower than 720p) options don't work, and anything above 720p (literally any 1080p~~, 2K DCI, 2160p, or 4K DCI setting) doesn't work either.

EDIT: Correction, after a restart 1080p60 in 8-bit YUV seems to work. Still unable to get 2160p59.94 output as I had hoped for.

It appears changing the resolution to a mode that doesn't work causes that output to break for the rest of the session; setting the resolution lower again never resumes output. Restarting FreeShow (with it set to a working resolution) 'frees it up' so to speak, allowing it to output once again. Maybe that might help diagnose? Also happy to provide logs if there are any.

I should also note that while the app seems to render very smoothly, Decklink output seems to be *very* laggy. Like, in 1080p60 mode, it's rendering transitions at maybe 3 fps.

In case it helps at all, here are my system specs:
Ryzen 5600X, 32GB DDR4 3200MT/s, ASUS ProArt B550-Creator, Samsung 980 500GB SSD, RTX A2000 12GB, Decklink 8K Pro, Intel X520 (10Gb SFP+ fiber)

**Comment by vassbo** [comment]
@Yttrium-tYcLief Thank you for the feedback! I think I got that fixed. (I updated the same link as before.)

**Comment by vassbo** [comment]
https://github.com/ChurchApps/FreeShow/releases/tag/v1.6.0-beta.1
